orb-ui/src/simulation.rs

```diff
@@ -1,90 +1,229 @@
+        let biometric_capture_interval_ms = 70;
+        for i in 1..=100 {
+            if !showcar && (30..=50).contains(&i) {
+                // simulate user moving away
+                ui.biometric_capture_distance(false);
+                ui.biometric_capture_occlusion(true);
+            } else {
+                // capture is making progress
+                ui.biometric_capture_distance(true);
+                ui.biometric_capture_occlusion(false);
+                ui.biometric_capture_distance(true);
+                ui.biometric_capture_progress(i as f64 / 100.0);
+            }
+
+            // randomly simulate error
+            if !showcar && i == 50 && rand::random::<u8>() % 5 == 0 {
+                info!("⚠️ Simulating biometric capture error");
+                biometric_capture_error = true;
+                ui.signup_fail(SignupFailReason::Timeout);
+                break;
+            }
+
+            // simulate gimbal movement
+            let x_base = if i < 50 { 40000 } else { 35000 };
+            let x_rand = rand::random::<i32>() % 1000;
+            let y_rand = rand::random::<i32>() % 1000;
+            let x_sign: i32 = if rand::random::<u8>() % 2 == 0 { 1 } else { -1 };
+            let y_sign: i32 = if rand::random::<u8>() % 2 == 0 { 1 } else { -1 };
+            x_angle = x_base + (x_rand * x_sign);
+            ui.gimbal(x_angle as u32, 90000 + (y_rand * y_sign) as u32);
+
+            time::sleep(Duration::from_millis(biometric_capture_interval_ms)).await;
+        }
+
+        if !biometric_capture_error {
+            // fill the ring
+            ui.biometric_capture_progress(1.1);
+            time::sleep(Duration::from_secs(1)).await;
+
+            ui.biometric_capture_success();
+
+            if showcar {
+                time::sleep(Duration::from_secs(4)).await;
+            }
+
+            if !self_serve {
+                // biometric pipeline, in 2 stages
+                // to test `starting_enrollment`
+                time::sleep(Duration::from_secs(1)).await;
+                for i in 0..=2 {
+                    ui.biometric_pipeline_progress(i as f64 / 5.0);
+                    time::sleep(Duration::from_secs(1)).await;
+                }
+                time::sleep(Duration::from_secs(1)).await;
+                ui.starting_enrollment();
+                time::sleep(Duration::from_secs(2)).await;
+                for i in 2..5 {
+                    ui.biometric_pipeline_progress(i as f64 / 5.0);
+                    time::sleep(Duration::from_millis(500)).await;
+                }
+                ui.biometric_pipeline_success();
+
+                time::sleep(Duration::from_secs(1)).await;
+                if rand::random::<u8>() % 2 == 0 {
+                    ui.signup_success();
+                } else {
+                    let fail_reason = SignupFailReason::from(
+                        rand::random::<u8>() % SignupFailReason::Unknown as u8,
```

Comment:
Maybe consider doing this: `https://www.reddit.com/r/rust/comments/i5n2zm/comment/g0rb9ob/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button`

The way its done right now feels a bit hacky, but if you're opinionated, we can keep it.

Reply:
it's quite straightforward for me but i've been writing C my whole life 😄

Reply:
done